src/llama.cpp

```diff
@@ -12670,7 +12687,17 @@ static void llama_set_inputs(llama_context & lctx, const llama_batch & batch) {
 
             GGML_ASSERT(ggml_backend_buffer_is_host(lctx.inp_KQ_mask->buffer));
 
-            float * data = (float *) lctx.inp_KQ_mask->data;
+            float * data     = (float *) lctx.inp_KQ_mask->data;
+            float * data_swa = nullptr;
+            const llama_pos n_keep_swa = hparams.n_sliding - batch.n_tokens;
```

Comment:
I don't understand the meaning of `n_keep_swa`. Seems this won't work with batches of multiple sequences

Reply:
Yeah I'm not sure if I'm doing it correctly: It is to emulate the rolling. If we input `n_tokens` then we only keep `n_sliding - n_tokens` tokens in cache, so the total number of tokens for attention is `n_tokens` plus `n_sliding - n_tokens` equals `n_sliding`

Reply:
Seems to me just restricting the position delta to be less than `n_swa` is enough:

```diff
diff --git a/src/llama.cpp b/src/llama.cpp
index 71b7ef62..fa207234 100644
--- a/src/llama.cpp
+++ b/src/llama.cpp
@@ -12722,7 +12722,7 @@ static void llama_set_inputs(llama_context & lctx, const llama_batch & batch) {
 
                         // may need to cut off old tokens for sliding window
                         if (data_swa) {
-                            if (pos - lctx.kv_self.cells[i].pos > n_keep_swa) {
+                            if (pos - lctx.kv_self.cells[i].pos >= hparams.n_sliding) {
                                 f = -INFINITY;
                             }
                             data_swa[h*(n_kv*n_tokens) + j*n_kv + i] = f;
```

This way, in SWA layers, the token with position 4096 does not "see" the token with position 0, but does "see" the token at position 1.

Reply:
OK thanks, that's clear for me now. I changed this code in https://github.com/ggerganov/llama.cpp/pull/8227/commits/ed5496fb32e9888abeaa7672aaba4d4251671457